uccb/src/bin/validator-oracle.rs

```diff
@@ -0,0 +1,273 @@
+use std::{path::PathBuf, str::FromStr};
+
+use alloy::{
+    contract::{ContractInstance, DynCallBuilder, Interface},
+    dyn_abi::DynSolValue,
+    eips::{eip2930::AccessList, BlockNumberOrTag},
+    json_abi::JsonAbi,
+    primitives::Address,
+    providers::Provider,
+    pubsub::PubSubFrontend,
+    rpc::types::Filter,
+    signers::local::PrivateKeySigner,
+};
+use anyhow::Result;
+use clap::Parser;
+use futures_util::stream::StreamExt;
+use tokio::sync::watch;
+use tracing::{debug, error, info};
+use tracing_subscriber::EnvFilter;
+use uccb::{
+    cfg::{ChainConfig, Config},
+    client::ChainClient,
+};
+use zilliqa::{contracts, crypto::SecretKey, state::contract_addr};
+
+const VALIDATOR_MANAGER_ABI_JSON: &str =
+    include_str!["../../contracts/out/ValidatorManager.sol/ValidatorManager.json"];
+
+#[derive(Parser, Debug)]
+struct Args {
+    #[arg(value_parser = SecretKey::from_hex)]
+    secret_key: SecretKey,
+    #[clap(long, short, default_value = "config.toml")]
+    config_file: PathBuf,
+}
+
+impl uccb::Args for Args {
+    fn secret_key(&self) -> &SecretKey {
+        &self.secret_key
+    }
+
+    fn config_file(&self) -> &PathBuf {
+        &self.config_file
+    }
+}
+
+// Workaround for displaying the collection of validators
+// using Display instead of Debug which shows too mube info...
+struct Display<'a>(&'a std::vec::Vec<Address>);
+
+impl<'a> std::fmt::Display for Display<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        if let Ok(s) = serde_json::to_string(self.0) {
+            write!(f, "{}", &s)?
+        }
+        Ok(())
+    }
+}
+
+struct ValidatorOracle {
+    signer: PrivateKeySigner,
+    chain_clients: Vec<ChainClient>,
+    deploy_abi: JsonAbi,
+    validator_manager_abi: JsonAbi,
+}
+
+impl ValidatorOracle {
+    pub async fn new(secret_key: SecretKey, config: Config) -> Result<Self> {
+        let signer = PrivateKeySigner::from_str(secret_key.to_hex().as_str())?;
+        let mut chain_clients = vec![Self::create_zq2_chain_client(config.clone(), &signer).await?];
+
+        for chain_config in config.chain_configs {
+            chain_clients.push(
+                ChainClient::new(
+                    &chain_config,
+                    config.zq2.validator_manager_address,
+                    signer.clone(),
+                )
+                .await?,
+            );
+        }
+
+        Ok(Self {
+            signer,
+            chain_clients,
+            deploy_abi: serde_json::from_value(serde_json::to_value(
+                contracts::deposit::ABI.clone(),
+            )?)?,
+            validator_manager_abi: serde_json::from_value(
+                serde_json::from_str::<serde_json::Value>(VALIDATOR_MANAGER_ABI_JSON)?["abi"]
+                    .clone(),
+            )?,
```

Comment:
We only need the abi field. and VALIDATOR_MANAGER_ABI_JSON is a &str

Reply:
It seems you are going through JSON because you need to convert a `ethabi::Contract` to a `alloy::json_abi::JsonAbi`. Can you add a comment explaining this and stating its valid because the JSON representation of both is the same - [Solidity ABI spec](https://docs.soliditylang.org/en/latest/abi-spec.html#json).